File: language/src/main/java/brj/nodes/LetNode.java
package brj.nodes;

import brj.BridjeLanguage;
import com.oracle.truffle.api.frame.VirtualFrame;
import com.oracle.truffle.api.nodes.ExplodeLoop;
import com.oracle.truffle.api.source.SourceSection;
import org.jetbrains.annotations.NotNull;

public class LetNode extends ExprNode {

    @Children
    private final WriteLocalNode[] bindingNodes;

    @Child
    private ExprNode expr;

    private final SourceSection sourceSection;

    public LetNode(BridjeLanguage lang, WriteLocalNode[] bindingNodes, ExprNode expr, SourceSection sourceSection) {
        super(lang);
        this.bindingNodes = bindingNodes;
        this.expr = expr;
        this.sourceSection = sourceSection;
    }

    @Override
    public SourceSection getSourceSection() {
        return sourceSection;
    }

    @NotNull
    @Override
    @ExplodeLoop
    public Object execute(@NotNull VirtualFrame frame) {
        for (WriteLocalNode bindingNode : bindingNodes) {
            bindingNode.execute(frame);
        }

        return expr.execute(frame);
    }
}

File: language/src/main/kotlin/brj/nodes/LetNode.kt
package brj.nodes

import brj.BridjeNode
import com.oracle.truffle.api.frame.VirtualFrame
import com.oracle.truffle.api.nodes.Node.Child

class LetNode(
    private val slot: Int,
    @field:Child private var bindingNode: BridjeNode,
    @field:Child private var bodyNode: BridjeNode
) : BridjeNode() {
    override fun execute(frame: VirtualFrame): Any? {
        val value = bindingNode.execute(frame)
        frame.setObject(slot, value)
        return bodyNode.execute(frame)
    }
}
